internal/configs/virtualserver_test.go

```diff
@@ -1361,6 +1361,357 @@ func TestGenerateVirtualServerConfigForVirtualServerWithMatches(t *testing.T) {
 	}
 }
 
+func TestGenerateVirtualServerConfigForVirtualServerRoutesWithDos(t *testing.T) {
+	dosResources := map[string]*appProtectDosResource{
+		"/coffee": {
+			AppProtectDosEnable:          "on",
+			AppProtectDosLogEnable:       false,
+			AppProtectDosMonitorUri:      "test.example.com",
+			AppProtectDosMonitorProtocol: "http",
+			AppProtectDosMonitorTimeout:  0,
+			AppProtectDosName:            "my-dos",
+			AppProtectDosAccessLogDst:    "svc.dns.com:123",
+			AppProtectDosPolicyFile:      "",
+			AppProtectDosLogConfFile:     "",
+		},
+	}
+
+	virtualServerEx := VirtualServerEx{
+		VirtualServer: &conf_v1.VirtualServer{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name:      "cafe",
+				Namespace: "default",
+			},
+			Spec: conf_v1.VirtualServerSpec{
+				Host: "cafe.example.com",
+				Upstreams: []conf_v1.Upstream{
+					{
+						Name:    "tea-v1",
+						Service: "tea-svc-v1",
+						Port:    80,
+					},
+					{
+						Name:    "tea-v2",
+						Service: "tea-svc-v2",
+						Port:    80,
+					},
+				},
+				Routes: []conf_v1.Route{
+					{
+						Path: "/tea",
+						Matches: []conf_v1.Match{
+							{
+								Conditions: []conf_v1.Condition{
+									{
+										Header: "x-version",
+										Value:  "v2",
+									},
+								},
+								Action: &conf_v1.Action{
+									Pass: "tea-v2",
+								},
+							},
+						},
+						Action: &conf_v1.Action{
+							Pass: "tea-v1",
+						},
+					},
+					{
+						Path:  "/coffee",
+						Route: "default/coffee",
+					},
+				},
+			},
+		},
+		Endpoints: map[string][]string{
+			"default/tea-svc-v1:80": {
+				"10.0.0.20:80",
+			},
+			"default/tea-svc-v2:80": {
+				"10.0.0.21:80",
+			},
+			"default/coffee-svc-v1:80": {
+				"10.0.0.30:80",
+			},
+			"default/coffee-svc-v2:80": {
+				"10.0.0.31:80",
+			},
+		},
+		VirtualServerRoutes: []*conf_v1.VirtualServerRoute{
+			{
+				ObjectMeta: meta_v1.ObjectMeta{
+					Name:      "coffee",
+					Namespace: "default",
+				},
+				Spec: conf_v1.VirtualServerRouteSpec{
+					Host: "cafe.example.com",
+					Upstreams: []conf_v1.Upstream{
+						{
+							Name:    "coffee-v1",
+							Service: "coffee-svc-v1",
+							Port:    80,
+						},
+						{
+							Name:    "coffee-v2",
+							Service: "coffee-svc-v2",
+							Port:    80,
+						},
+					},
+					Subroutes: []conf_v1.Route{
+						{
+							Path: "/coffee",
```

Comment:
because the new code calls `addDosConfigToLocations` for 3 different cases, I suggest testing them all. For that, we can have a VSR with 3 subroutes:
- one a regular around with a `pass` or `proxy` action 
- one subroute with splits
- one subroute with matches (the one that we have now)